File: 0383-ransom-note/0383-ransom-note.js
/**
 * @param {string} ransomNote
 * @param {string} magazine
 * @return {boolean}
 */
var canConstruct = function(ransomNote, magazine) {
    const freq = {};
    for (let letter of magazine) {
        freq[letter] = (freq[letter] || 0) + 1;
    }
    
    for (let letter of ransomNote) {
        if (!freq[letter]) return false;
        freq[letter]--;
    }
    
    return true;
};

File: 0383-ransom-note/0383-ransom-note.py
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        freq = {}
        for letter in magazine:
            if letter in freq:
                freq[letter] += 1
            else:
                freq[letter] = 1
        
        for letter in ransomNote:
            if letter not in freq:
                return False
            else:
                freq[letter] -= 1
                if freq[letter] < 0:
                    return False
        
        return True
        